clustered/common/src/main/java/org/ehcache/clustered/common/messages/ChainCodec.java

```diff
@@ -0,0 +1,150 @@
+        msg = combine(buffer, encodeElement(element));
+        continue;
+      }
+      if (msg == null) {
+        throw new IllegalArgumentException("Message cannot be null");
+      }
+      msg = combine(msg, encodeElement(element));
+
+    }
+    return msg != null ? msg.array() : new byte[0];
+  }
+
+  public static Chain decode(byte[] payload) {
+    final List<Element> elements = new ArrayList<Element>();
+    if (payload.length != 0) {
+      ByteBuffer buffer = ByteBuffer.wrap(payload);
+      boolean isSequenced = buffer.get() == 1;
+      if (isSequenced) {
+        while (buffer.hasRemaining()) {
+          final long sequence = buffer.getLong();
+          int payloadSize = buffer.getInt();
+          buffer.limit(buffer.position() + payloadSize);
+          final ByteBuffer elementPayload = buffer.slice();
+          buffer.position(buffer.limit());
+          buffer.limit(buffer.capacity());
+          elements.add(new SequencedElement() {
+            @Override
+            public long getSequenceNumber() {
+              return sequence;
+            }
+
+            @Override
+            public ByteBuffer getPayload() {
+              return elementPayload.duplicate();
+            }
+          });
+        }
+      } else {
+        while (buffer.hasRemaining()) {
+          int payloadSize = buffer.getInt();
+          buffer.limit(buffer.position() + payloadSize);
+          final ByteBuffer elementPayload = buffer.slice();
+          buffer.position(buffer.limit());
+          buffer.limit(buffer.capacity());
+          elements.add(new Element() {
+            @Override
+            public ByteBuffer getPayload() {
+              return elementPayload.duplicate();
+            }
+          });
+        }
+      }
+    }
+
+    return new Chain() {
+
+      final List<Element> elementList = Collections.unmodifiableList(elements);
+
+      @Override
+      public Iterator<Element> reverseIterator() {
+        return Util.reverseIterator(elementList);
+      }
+
+      @Override
+      public boolean isEmpty() {
+        return elementList.isEmpty();
+      }
+
+      @Override
+      public Iterator<Element> iterator() {
+        return elementList.iterator();
+      }
+    };
+  }
+
+  private static ByteBuffer combine(ByteBuffer buffer1, ByteBuffer buffer2) {
+    ByteBuffer byteBuffer = ByteBuffer.allocate(buffer1.remaining() + buffer2.remaining());
+    byteBuffer.put(buffer1);
+    byteBuffer.put(buffer2);
+    byteBuffer.flip();
+    return byteBuffer;
+  }
+
+  private static ByteBuffer encodeElement(Element element) {
+    if (element instanceof SequencedElement) {
+      ByteBuffer buffer = ByteBuffer.allocate(SEQ_NUM_OFFSET + ELEMENT_PAYLOAD_OFFSET + element.getPayload().remaining());
+      buffer.putLong(((SequencedElement)element).getSequenceNumber());
+      buffer.putInt(element.getPayload().remaining());
+      buffer.put(element.getPayload());
+      buffer.flip();
+      return buffer;
+    }
```

Comment:
else could be better here..also some duplicate code can be avoided (from 139-142) if that part of the code is made common. Can we also abstract the allocate difference between sequenced and normal elements and use a template pattern here?. Again possibly a future refactoring.